core/validator/src/rules/duplicate-name.ts

```diff
@@ -47,37 +48,48 @@ export class DuplicateName extends BaseValidator {
    */
   Logic(
     metadatas: Metadata[],
-    validations: Record<string, StateOutput>,
+    validations: Record<string, StateOutput>
   ): Record<string, StateOutput> {
-    const seen = {
-      names: new Set<string>(),
-    };
+    const errorsMetadata = new Set<Metadata>();
+    const nameCount: Record<string, number> = {};
 
+    // First pass: Count occurrences of each name
     for (const entry of metadatas) {
-      if (
-        typeof entry === "object" &&
-        entry !== null &&
-        "name" in entry.metadata
-      ) {
-        if (seen.names.has(entry.metadata.name)) {
-          if (!validations[entry.assetName]) {
-            validations[entry.assetName] = {
-              status: "error",
-              warnings: [],
-              errors: [],
-            };
-          }
+      const name = entry.metadata.name;
+      nameCount[name] = (nameCount[name] || 0) + 1;
+    }
 
-          validations[entry.assetName].status = "error";
-          validations[entry.assetName].errors.push({
-            validatorId: this.id,
-            message: `Name: ${entry.metadata.name} has been detected as a duplicate.`,
-          });
-        }
-        seen.names.add(entry.metadata.name);
+    // Second pass: Identify duplicates based on the count
+    for (const entry of metadatas) {
+      const name = entry.metadata.name;
+      if (nameCount[name] > 1) {
+        errorsMetadata.add(entry);
```

Comment:
same as above, doing both check in the same loop as the data is already up to date